feature: added support for storing generated HTML report in a variable

sankha555 opened the pull request:
* [x] fix #144 
* [x] added support for storing generated HTML report in a variable which can be returned from the ```create_report()``` function
* [x] tests added and passed

sankha555 commented:
@matthiaskoenig I have made a patch to enable returning the HTML report generated as a variable. I have implemented it in the following manner:

- added a boolean ```report_variable``` parameter in ```create_reports()``` and ```create_report()``` to specify if report has to be returned as a variable (i.e. ```report_variable = True```)
- Based on the value of the ```report_variable```, the report would either return the HTML as a variable or proceed to write the HTML report to a file
- Added a test ```test_report_variable``` in ```test_report.py``` to generate the report two times, once with ```report_variable = True``` and once with ```report_variable = False```. The report file is read and its contents are matched with the returned variable's data to assert if both the times, the report was generated properly.

For the ```create_reports()``` function, I approached the task by concatenating the content of the index HTML file with the content of the report for each of the model in the ```sbml_paths``` list. This concatenated string is returned as the variable then. I am a bit confused if this approach would be useful to us, do you think we can improve it in any manner? 

I was also thinking if we should add a new function altogether to return the HTML variable (something like ```get_html_variable()```) and use the validations of ```create_report()``` in it. Which of these two approaches would be more useful and efficient in your opinion?

Please let me know about your feedback on the pull request and if I should make any changes to improvise it. Thanks!

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -39,6 +39,7 @@ def create_reports(
     promote: bool = False,
     validate: bool = True,
     math_type: str = "cmathml",
+    report_variable: bool = False,
```
Do not add a variable. Just let the function always return the report string. You can add the functionality of returning the string without any additional arguments.
Annotate return type and update description.

matthiaskoenig replied:
This function must return a list of report strings, i.e. `List[str]`

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -61,19 +64,31 @@ def create_reports(
             promote=promote,
             validate=validate,
             math_type=math_type,
+            report_variable=report_variable,
```
not needed

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -89,6 +104,7 @@ def create_report(
     log_errors: bool = True,
     units_consistency: bool = True,
     modeling_practice: bool = True,
+    report_variable: bool = False,
```
not needed

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -105,8 +121,9 @@ def create_report(
+    :param report_variable: boolean flag to specify storing HTML report in a variable
 
-    :return:
+    :return: html variable containing report HTML content if report_variable is True
```
always return the html string

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -105,8 +121,9 @@ def create_report(
     :param log_errors: boolean flag of errors should be logged
     :param units_consistency: boolean flag units consistency
     :param modeling_practice: boolean flag modeling practise
+    :param report_variable: boolean flag to specify storing HTML report in a variable
```
not needed

matthiaskoenig commented on src/sbmlutils/report/sbmlreport.py:
```diff
@@ -151,8 +168,13 @@ def create_report(
     html = _create_html(doc, basename, html_template=template, math_type=math_type)
+
+    # store and return the variable if HTML report is to be stored in a variable
+    if report_variable is True:
```
always return html

matthiaskoenig commented on src/sbmlutils/test/report/test_report.py:
```diff
@@ -72,3 +72,29 @@ def check_report_math_type(sbml_path: Path, math_type: str, tmp_path):
+@pytest.mark.parametrize("sbml_path", sbml_paths, ids=sbml_paths_idfn)
+def test_report_variable(sbml_path, tmp_path):
+    """Test report generation."""
+    html = sbmlreport.create_report(
```
you only need a single test

matthiaskoenig commented on src/sbmlutils/test/report/test_report.py:
```diff
@@ -72,3 +72,29 @@ def check_report_math_type(sbml_path: Path, math_type: str, tmp_path):
+        output_dir=tmp_path,
+        report_variable=True,
+    )
+
```
add asserts:
```
assert html
assert isinstance(html, str)
```

matthiaskoenig commented:
Simplify the approach:
You should always return the html string without any need for an argument.
The create_reports should return the List[str] of the individual html reports (index.html is not important here).

sankha555 commented:
I understand the requirements now, I had previously thought that the variable return feature should be added as an additional functionality apart from writing the HTML file. Sure, I will make these changes in the corresponding functions and remove the code for writing the HTML to the files. I will update the pull request in a while. Thanks!

sankha555 commented:
@matthiaskoenig I have made the changes to the , please take a look. 
In the tests, I have removed the code for checking the existence of the HTML report file in the ```check_report_math_type()``` and replaced it with the assertion for the HTML variable. Would this change be suitable for catering to both the purposes (checking the math type as well as the report variable in one test)?

Also, I have removed the code for generating the ```index.html``` file in the ```create_reports()``` function, but I haven't removed the function ```_create_index_html()``` in case we need it in future. If I should make any amendments to these changes, please let me know. I would make the updates and push the changes for review. Thanks!
